### PR TITLE
utils refactor clean-up

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Few follow-up items after #1421 was completed. Some helpers were in the wrong locatino to avoid circular dependencies. We moved those back and remove some pesky dependencies between modules that pollute our dependency graph.
- Removes the config <> training dependency. config should not depend on anything in training, only recipes should depend on training. Instead of checking for rank zero without `log_config`, do it in distribtued recipes instead.
- Moves `_get_prompt_template` to `data/` so `config/` does not depend on any other directory other than `utils/`
- Moves `deprecated` from `data/` to `utils/`, a more appropriate location.

#### Changelog
What are the changes made in this PR?

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

File: torchtune/config/_utils.py [hunk 1]
```diff
@@ -12,28 +12,17 @@
...
     """
-    Logs the resolved config (merged YAML file and CLI overrides) to rank zero.
+    Logs the resolved config (merged YAML file and CLI overrides).
 
     Args:
         recipe_name (str): name of the recipe to display
         cfg (DictConfig): parsed config object
     """
-    # Log the config only on rank 0
-    _, rank = get_world_size_and_rank()
```
Comment (ebsmothers) [inline torchtune/config/_utils.py:33]: rank = torch.distributed.get_rank() if torch.distributed.is_available() and torch.distributed.is_initialized() else 0

**Review by felipemello1 — APPROVED** [review]

File: recipes/dev/lora_finetune_fsdp2.py [hunk 2]
```diff
@@ -718,9 +718,7 @@ def recipe_main(cfg: DictConfig) -> None:
     os.environ["TORCH_NCCL_AVOID_RECORD_STREAMS"] = "1"
     init_process_group(backend="gloo" if cfg.device == "cpu" else "nccl")
 
-    _, rank = training.get_world_size_and_rank()
-    if rank == 0:
-        config.log_config(recipe_name="LoRAFinetuneRecipeDistributed", cfg=cfg)
+    config.log_config(recipe_name="LoRAFinetuneRecipeDistributed", cfg=cfg)
```
Comment (felipemello1) [inline recipes/dev/lora_finetune_fsdp2.py]: i think you missed adding the rank here

File: tests/torchtune/config/test_config_utils.py [hunk 3]
```diff
@@ -128,12 +128,37 @@ def test_log_config(self, capsys):
...
+                )
+
+            # Clear the stream
+            stream.truncate(0)
+            stream.seek(0)
+
+            # Make sure all other ranks do not log anything
+            with mock.patch(
+                "torchtune.config._utils.dist.get_rank",
+                return_value=1,
```
Comment (felipemello1) [inline tests/torchtune/config/test_config_utils.py:157]: i thought that you removed it because logging doesnt check for rank anymore

**Review by ebsmothers — APPROVED** [review]
